File: api-gateway/src/modules/professor/professor.controller.ts
import { Controller, Get, Post, Put, Param, Body, Headers, Query } from '@nestjs/common';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import sanitizeHtml from 'sanitize-html';
import { ProfessorService } from './professor.service';
import {
    CreateProfessorProfileDto, UpdateProfessorProfileDto,
    CreateExamDto, UpdateExamDto, CreateExamQuestionDto,
    SubmitExamAttemptDto, GradeOpenTextDto,
    CreateCertificateTemplateDto, UpdateCertificateTemplateDto, ManualIssueCertificateDto,
} from './professor.dto';

@ApiTags('Professor')
@ApiBearerAuth()
@Controller('professor')
export class ProfessorController {
    constructor(private readonly professorService: ProfessorService) {}

    // ── Profile ───────────────────────────────────────────────────────────────
    @Post('profile')
    createProfile(@Body() dto: CreateProfessorProfileDto, @Headers('authorization') token: string) {
        return this.professorService.createProfile(dto, token);
    }

    @Get('profile')
    getMyProfile(@Headers('authorization') token: string) {
        return this.professorService.getMyProfile(token);
    }

    @Put('profile')
    updateProfile(@Body() dto: UpdateProfessorProfileDto, @Headers('authorization') token: string) {
        return this.professorService.updateProfile(dto, token);
    }

    // ── Dashboard ─────────────────────────────────────────────────────────────
    @Get('dashboard')
    getDashboard(@Headers('authorization') token: string) {
        return this.professorService.getDashboard(token);
    }

    @Get('earnings')
    getEarnings(@Headers('authorization') token: string) {
        return this.professorService.getEarnings(token);
    }

    @Get('students')
    getStudents(@Headers('authorization') token: string) {
        return this.professorService.getStudents(token);
    }

    // ── Stripe Connect ────────────────────────────────────────────────────────
    @Post('stripe/connect')
    createStripeConnect(@Query('returnUrl') returnUrl: string, @Headers('authorization') token: string) {
        return this.professorService.createStripeConnect(returnUrl, token);
    }

    @Get('stripe/status')
    getStripeStatus(@Headers('authorization') token: string) {
        return this.professorService.getStripeStatus(token);
    }

    // ── Exams ─────────────────────────────────────────────────────────────────
    @Post('exams')
    createExam(@Body() dto: CreateExamDto, @Headers('authorization') token: string) {
        return this.professorService.createExam(dto, token);
    }

    @Get('exams/:examId')
    getExam(@Param('examId') examId: string, @Headers('authorization') token: string) {
        return this.professorService.getExam(examId, token);
    }

    @Get('exams/course/:courseId')
    getExamsByCourse(@Param('courseId') courseId: string, @Headers('authorization') token: string) {
        return this.professorService.getExamsByCourse(courseId, token);
    }

    @Put('exams/:examId')
    updateExam(@Param('examId') examId: string, @Body() dto: UpdateExamDto, @Headers('authorization') token: string) {
        return this.professorService.updateExam(examId, dto, token);
    }

    @Post('exams/:examId/publish')
    publishExam(@Param('examId') examId: string, @Headers('authorization') token: string) {
        return this.professorService.publishExam(examId, token);
    }

    @Post('exams/:examId/questions')
    addQuestion(@Param('examId') examId: string, @Body() dto: CreateExamQuestionDto, @Headers('authorization') token: string) {
        return this.professorService.addQuestion(examId, dto, token);
    }

    @Get('exams/:examId/attempts')
    getAttempts(@Param('examId') examId: string, @Headers('authorization') token: string) {
        return this.professorService.getAttempts(examId, token);
    }

    @Post('exams/submit')
    submitAttempt(@Body() dto: SubmitExamAttemptDto, @Headers('authorization') token: string) {
        return this.professorService.submitAttempt(dto, token);
    }

    @Post('exams/attempts/grade')
    gradeOpenText(@Body() dto: GradeOpenTextDto, @Headers('authorization') token: string) {
        return this.professorService.gradeOpenText(dto, token);
    }

    // ── Courses ───────────────────────────────────────────────────────────────
    @Get('courses')
    getMyCourses(@Headers('authorization') token: string) {
        return this.professorService.getMyCourses(token);
    }

    @Post('courses')
    createCourse(@Body() dto: any, @Headers('authorization') token: string) {
        return this.professorService.createCourse(dto, token);
    }

    @Post('courses/:courseId/publish')
    publishCourse(@Param('courseId') courseId: string, @Headers('authorization') token: string) {
        return this.professorService.publishCourse(courseId, token);
    }

    // ── Certificates ──────────────────────────────────────────────────────────
    @Get('certificates/templates')
    getCertTemplates(@Headers('authorization') token: string) {
        return this.professorService.getCertificateTemplates(token);
    }

    @Post('certificates/templates')
    createCertTemplate(@Body() dto: CreateCertificateTemplateDto, @Headers('authorization') token: string) {
        if (dto.bodyHtml) {
            dto.bodyHtml = sanitizeHtml(dto.bodyHtml, {
                allowedTags: sanitizeHtml.defaults.allowedTags.concat(['h1', 'h2', 'img']),
                allowedAttributes: {
                    ...sanitizeHtml.defaults.allowedAttributes,
                    '*': ['style', 'class']
                }
            });
        }
        return this.professorService.createCertificateTemplate(dto, token);
    }

    @Put('certificates/templates/:templateId')
    updateCertTemplate(@Param('templateId') id: string, @Body() dto: UpdateCertificateTemplateDto, @Headers('authorization') token: string) {
        if (dto.bodyHtml) {
            dto.bodyHtml = sanitizeHtml(dto.bodyHtml, {
                allowedTags: sanitizeHtml.defaults.allowedTags.concat(['h1', 'h2', 'img']),
                allowedAttributes: {
                    ...sanitizeHtml.defaults.allowedAttributes,
                    '*': ['style', 'class']
                }
            });
        }
        return this.professorService.updateCertificateTemplate(id, dto, token);
    }

    @Post('certificates/issue')
    manualIssue(@Body() dto: ManualIssueCertificateDto, @Headers('authorization') token: string) {
        return this.professorService.manualIssueCertificate(dto, token);
    }

    @Get('certificates/course/:courseId')
    getCertsByCourse(@Param('courseId') courseId: string, @Headers('authorization') token: string) {
        return this.professorService.getCertsByCourse(courseId, token);
    }

    // ── Live Streams ──────────────────────────────────────────────────────────
    @Get('livestreams')
    getLiveStreams(@Headers('authorization') token: string) {
        return this.professorService.getLiveStreams(token);
    }

    @Get('livestreams/youtube-status')
    getYouTubeStatus(@Headers('authorization') token: string) {
        return this.professorService.getYouTubeStatus(token);
    }

    @Get('livestreams/oauth/url')
    getYouTubeOAuthUrl(@Headers('authorization') token: string) {
        return this.professorService.getYouTubeOAuthUrl(token);
    }

    @Post('livestreams/oauth/exchange')
    exchangeYouTubeCode(@Body() body: { code: string }, @Headers('authorization') token: string) {
        return this.professorService.exchangeYouTubeCode(body, token);
    }

    @Get('livestreams/course/:courseId/active')
    getActiveCourseSession(@Param('courseId') courseId: string, @Headers('authorization') token: string) {
        return this.professorService.getActiveCourseSession(courseId, token);
    }

    @Get('livestreams/:id/stream-key')
    getStreamKey(@Param('id') id: string, @Headers('authorization') token: string) {
        return this.professorService.getStreamKey(id, token);
    }

    @Get('livestreams/:id')
    getLiveStream(@Param('id') id: string, @Headers('authorization') token: string) {
        return this.professorService.getLiveStream(id, token);
    }

    @Post('livestreams')
    createLiveStream(@Body() body: any, @Headers('authorization') token: string) {
        return this.professorService.createLiveStream(body, token);
    }

    @Post('livestreams/:id/start')
    startLiveStream(@Param('id') id: string, @Headers('authorization') token: string) {
        return this.professorService.startLiveStream(id, token);
    }

    @Post('livestreams/:id/end')
    endLiveStream(@Param('id') id: string, @Headers('authorization') token: string) {
        return this.professorService.endLiveStream(id, token);
    }
}

File: Core/SkillmindCore/SkillMind.WebAPI/Controllers/v1/ProfessorController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;
using SkillMind.Core.Application.Dtos.Professor;
using SkillMind.Core.Application.Dtos.Courses;
using SkillMind.Core.Application.Interfaces;
using SkillMind.Core.Domain.Enums;
using SkillMind.Infrastructure.Shared.Services;

namespace SkillMind.WebAPI.Controllers.v1;

[Authorize(Roles = "Professor,Admin")]
public class ProfessorController(
    IProfessorService professorService,
    IExamService examService,
    ICertificateService certificateService,
    ICourseService courseService,
    StripeServices stripeServices) : BaseController
{
    // ── Profile ───────────────────────────────────────────────────────────────

    [HttpPost("profile")]
    [AllowAnonymous] // Called right after email verification, before auth token is stable
    [ProducesResponseType(typeof(ProfessorProfileDto), StatusCodes.Status201Created)]
    [ProducesResponseType(typeof(ProfessorProfileDto), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> CreateProfile([FromBody] CreateProfessorProfileDto dto)
    {
        try
        {
            var existing = await professorService.GetProfileByUserIdAsync(dto.UserId);
            if (existing is not null)
                return Ok(existing);

            var result = await professorService.CreateProfileAsync(dto);
            return Created(string.Empty, result);
        }
        catch (InvalidOperationException ex) { return BadRequest(new { Message = ex.Message }); }
    }

    [HttpGet("profile")]
    [ProducesResponseType(typeof(ProfessorProfileDto), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetMyProfile()
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (string.IsNullOrEmpty(userId)) return Unauthorized();

        var profile = await professorService.GetProfileByUserIdAsync(userId);
        return profile is null ? NotFound() : Ok(profile);
    }

    [HttpPut("profile")]
    [ProducesResponseType(typeof(ProfessorProfileDto), StatusCodes.Status200OK)]
    public async Task<IActionResult> UpdateProfile([FromBody] UpdateProfessorProfileDto dto)
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (string.IsNullOrEmpty(userId)) return Unauthorized();

        var profile = await professorService.GetProfileByUserIdAsync(userId);
        if (profile is null) return NotFound("Professor profile not found.");

        var result = await professorService.UpdateProfileAsync(profile.Id, dto);
        return Ok(result);
    }

    // ── Dashboard ─────────────────────────────────────────────────────────────

    [HttpGet("dashboard")]
    [ProducesResponseType(typeof(ProfessorDashboardDto), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetDashboard()
    {
        var profile = await GetCurrentProfileAsync();
        if (profile is null) return NotFound("Professor profile not found.");

        var dashboard = await professorService.GetDashboardAsync(profile.Id);
        return Ok(dashboard);
    }

    [HttpGet("students")]
    [ProducesResponseType(typeof(List<EnrolledStudentDto>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetEnrolledStudents()
    {
        var profile = await GetCurrentProfileAsync();
        if (profile is null) return NotFound("Professor profile not found.");

        var students = await professorService.GetEnrolledStudentsAsync(profile.Id);
        return Ok(students);
    }

    // ── Stripe Connect ────────────────────────────────────────────────────────

    [HttpPost("stripe/connect")]
    [ProducesResponseType(typeof(StripeConnectOnboardingDto), StatusCodes.Status200OK)]
    public async Task<IActionResult> CreateStripeConnect([FromQuery] string returnUrl)
    {
        var profile = await GetCurrentProfileAsync();
        if (profile is null) return NotFound("Professor profile not found.");

        try
        {
            var refreshUrl = returnUrl + "?refresh=1";
            var (accountId, onboardingUrl) = await stripeServices.CreateConnectAccountAsync(
                profile.HasStripeConnect ? await professorService.GetStripeAccountIdAsync(profile.Id) : null,
                returnUrl,
                refreshUrl);

            await professorService.SaveStripeAccountAsync(profile.Id, accountId, PayoutStatus.Pending);

            return Ok(new StripeConnectOnboardingDto { OnboardingUrl = onboardingUrl });
        }
        catch (Stripe.StripeException ex)
        {
            return BadRequest(new { Message = ex.Message });
        }
    }

    [HttpGet("stripe/status")]
    [ProducesResponseType(typeof(StripeConnectStatusDto), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetStripeStatus()
    {
        var profile = await GetCurrentProfileAsync();
        if (profile is null) return NotFound("Professor profile not found.");

        var accountId = await professorService.GetStripeAccountIdAsync(profile.Id);
        if (string.IsNullOrEmpty(accountId))
        {
            return Ok(new StripeConnectStatusDto
            {
                PayoutStatus = PayoutStatus.NotConfigured.ToString(),
                ChargesEnabled = false,
                PayoutsEnabled = false,
                StripeAccountId = null
            });
        }

        try
        {
            var (chargesEnabled, payoutsEnabled) = await stripeServices.GetConnectAccountStatusAsync(accountId);
            var status = chargesEnabled && payoutsEnabled ? PayoutStatus.Active
                : accountId is not null ? PayoutStatus.Pending
                : PayoutStatus.NotConfigured;

            // Persist updated status
            await professorService.SaveStripeAccountAsync(profile.Id, accountId, status);

            return Ok(new StripeConnectStatusDto
            {
                PayoutStatus = status.ToString(),
                ChargesEnabled = chargesEnabled,
                PayoutsEnabled = payoutsEnabled,
                StripeAccountId = accountId
            });
        }
        catch (Stripe.StripeException ex)
        {
            return BadRequest(new { Message = ex.Message });
        }
    }

    [HttpGet("earnings")]
    [ProducesResponseType(typeof(EarningsSummaryDto), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetEarningsFull()
    {
        var profile = await GetCurrentProfileAsync();
        if (profile is null) return NotFound("Professor profile not found.");

        var earnings = await professorService.GetEarningsSummaryAsync(profile.Id);

        // Enrich with real pending payout from Stripe if connected
        var accountId = await professorService.GetStripeAccountIdAsync(profile.Id);
        if (!string.IsNullOrEmpty(accountId))
        {
            try
            {
                earnings.PendingPayout = await stripeServices.GetConnectPendingBalanceAsync(accountId);
            }
            catch (Stripe.StripeException)
            {
                // Non-fatal: keep 0 if Stripe call fails
            }
        }

        return Ok(earnings);
    }

    [HttpPost("stripe/webhook")]
    [AllowAnonymous]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> StripeConnectWebhook()
    {
        var stripeSignature = Request.Headers["Stripe-Signature"].FirstOrDefault();
        if (string.IsNullOrWhiteSpace(stripeSignature))
            return BadRequest("Missing Stripe-Signature header.");

        using var reader = new StreamReader(Request.Body);
        var payload = await reader.ReadToEndAsync();

        // Stripe v2 thin events (type starts with "v2.") use a different format
        // that the Stripe.net SDK cannot parse with ConstructEvent — acknowledge and skip.
        if (payload.Contains("\"v2."))
            return Ok();

        try
        {
            var stripeEvent = stripeServices.ConstructConnectEvent(payload, stripeSignature);

            if (stripeEvent.Type == "account.updated")
            {
                var account = stripeEvent.Data.Object as Stripe.Account;
                if (account is not null)
                    await professorService.SyncStripeConnectStatusAsync(account.Id);
            }

            return Ok();
        }
        catch (Stripe.StripeException ex)
        {
            return BadRequest(new { Message = ex.Message });
        }
    }

    // ── Certificate Templates ─────────────────────────────────────────────────

    [HttpGet("certificates/templates")]
    [ProducesResponseType(typeof(List<CertificateTemplateDto>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetTemplates()
    {
        var profile = await GetCurrentProfileAsync();
        if (profile is null) return NotFound();

        var templates = await certificateService.GetTemplatesByProfessorAsync(profile.Id);
        return Ok(templates);
    }

    [HttpPost("certificates/templates")]
    [ProducesResponseType(typeof(CertificateTemplateDto), StatusCodes.Status201Created)]
    public async Task<IActionResult> CreateTemplate([FromBody] CreateCertificateTemplateDto dto)
    {
        var profile = await GetCurrentProfileAsync();
        if (profile is null) return NotFound();

        dto = dto with { ProfessorId = profile.Id };
        var result = await certificateService.CreateTemplateAsync(dto);
        return Created(string.Empty, result);
    }

    [HttpPut("certificates/templates/{templateId:guid}")]
    [ProducesResponseType(typeof(CertificateTemplateDto), StatusCodes.Status200OK)]
    public async Task<IActionResult> UpdateTemplate([FromRoute] Guid templateId, [FromBody] UpdateCertificateTemplateDto dto)
    {
        var result = await certificateService.UpdateTemplateAsync(templateId, dto);
        return Ok(result);
    }

    [HttpPost("certificates/issue")]
    [ProducesResponseType(typeof(CertificateDto), StatusCodes.Status201Created)]
    public async Task<IActionResult> ManualIssueCertificate([FromBody] ManualIssueCertificateDto dto)
    {
        try
        {
            var result = await certificateService.ManualIssueAsync(dto);
            return Created(string.Empty, result);
        }
        catch (InvalidOperationException ex) { return Conflict(new { Message = ex.Message }); }
    }

    [HttpGet("certificates/course/{courseId:guid}")]
    [ProducesResponseType(typeof(List<CertificateDto>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetCertsByCourse([FromRoute] Guid courseId)
    {
        var certs = await certificateService.GetCertificatesByCourseAsync(courseId);
        return Ok(certs);
    }

    [HttpGet("certificates/all")]
    [ProducesResponseType(typeof(List<CertificateDto>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetAllMyCerts()
    {
        var profile = await GetCurrentProfileAsync();
        if (profile is null) return NotFound();

        var certs = await certificateService.GetAllCertificatesByProfessorAsync(profile.Id);
        return Ok(certs);
    }

    // ── Exams ────────────────────────────────────────────────────────────────

    [HttpPost("exams")]
    [ProducesResponseType(typeof(ExamDto), StatusCodes.Status201Created)]
    public async Task<IActionResult> CreateExam([FromBody] CreateExamDto dto)
    {
        var result = await examService.CreateExamAsync(dto);
        return Created(string.Empty, result);
    }

    [HttpGet("exams/course/{courseId:guid}")]
    [ProducesResponseType(typeof(List<ExamDto>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetExamsByCourse([FromRoute] Guid courseId)
    {
        var result = await examService.GetExamsByCourseAsync(courseId);
        return Ok(result);
    }

    [HttpGet("exams/{examId:guid}")]
    [ProducesResponseType(typeof(ExamDto), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetExam([FromRoute] Guid examId)
    {
        var result = await examService.GetExamAsync(examId);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPut("exams/{examId:guid}")]
    [ProducesResponseType(typeof(ExamDto), StatusCodes.Status200OK)]
    public async Task<IActionResult> UpdateExam([FromRoute] Guid examId, [FromBody] UpdateExamDto dto)
    {
        var result = await examService.UpdateExamAsync(examId, dto);
        return Ok(result);
    }

    [HttpPost("exams/{examId:guid}/publish")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<IActionResult> PublishExam([FromRoute] Guid examId)
    {
        await examService.PublishExamAsync(examId);
        return NoContent();
    }

    [HttpPost("exams/{examId:guid}/questions")]
    [ProducesResponseType(typeof(ExamDto), StatusCodes.Status200OK)]
    public async Task<IActionResult> AddQuestion([FromRoute] Guid examId, [FromBody] CreateExamQuestionDto dto)
    {
        dto = dto with { ExamId = examId };
        var result = await examService.AddQuestionAsync(dto);
        return Ok(result);
    }

    [HttpGet("exams/{examId:guid}/attempts")]
    [ProducesResponseType(typeof(List<ExamAttemptDto>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetAttempts([FromRoute] Guid examId)
    {
        var result = await examService.GetAttemptsByExamAsync(examId);
        return Ok(result);
    }

    [HttpPost("exams/attempts/grade")]
    [ProducesResponseType(typeof(ExamAttemptDto), StatusCodes.Status200OK)]
    public async Task<IActionResult> GradeOpenText([FromBody] GradeOpenTextDto dto)
    {
        var result = await examService.GradeOpenTextAnswerAsync(dto);
        return Ok(result);
    }

    // ── Courses ───────────────────────────────────────────────────────────────

    [HttpGet("courses")]
    [ProducesResponseType(typeof(List<CourseDto>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetMyCourses()
    {
        var profile = await GetCurrentProfileAsync();
        if (profile is null) return NotFound("Professor profile not found.");

        var courses = await courseService.GetByProfessorAsync(profile.Id);
        return Ok(courses);
    }

    [HttpPost("courses")]
    [ProducesResponseType(typeof(CourseDto), StatusCodes.Status201Created)]
    public async Task<IActionResult> CreateCourse([FromBody] CreateCourseDto dto)
    {
        var profile = await GetCurrentProfileAsync();
        if (profile is null) return NotFound("Professor profile not found.");

        var course = await courseService.CreateCourseForProfessorAsync(dto, profile.Id);
        return Created(string.Empty, course);
    }

    [HttpPost("courses/{courseId:guid}/publish")]
    [ProducesResponseType(typeof(CourseDto), StatusCodes.Status200OK)]
    public async Task<IActionResult> PublishCourse(Guid courseId)
    {
        var profile = await GetCurrentProfileAsync();
        if (profile is null) return NotFound("Professor profile not found.");

        try
        {
            var course = await courseService.PublishCourseAsync(courseId, profile.Id);
            return Ok(course);
        }
        catch (UnauthorizedAccessException ex)
        {
            return Forbid();
        }
    }

    // ── Private helpers ───────────────────────────────────────────────────────

    private async Task<ProfessorProfileDto?> GetCurrentProfileAsync()
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (string.IsNullOrEmpty(userId)) return null;
        return await professorService.GetProfileByUserIdAsync(userId);
    }
}
